omnigibson/transition_rules.py

```diff
@@ -615,39 +743,35 @@ def transition(self, individual_objects, group_objects):
                 t_results.remove.append(obj)
 
         # Remove all physical particles that are inside the blender
-        for system_name in self.particle_requirements.keys():
+        for system_name in cls.particle_requirements.keys():
             system = get_system(system_name)
             # No need to check for whether particle instancers exist because they must due to @self.condition passing!
             for inst in system.particle_instancers.values():
-                indices = self._check_in_volume[blender.name](inst.particle_positions).nonzero()[0]
+                indices = cls._CHECK_IN_VOLUME[blender.name](inst.particle_positions).nonzero()[0]
                 inst.remove_particles(idxs=indices)
 
         # Spawn in blended physical particles!
-        blender.states[Filled].set_value(get_system(self.output_system), True)
+        blender.states[Filled].set_value(get_system(cls.output_system), True)
```

Comment:
Sanjana, Sharon and I had a brief discussion about this. Four modifications are probably needed in the BlenderRule.
1. BDDL will not provide the quantity of substance (e.g. milk) in the blender rule. So we should essentially check for `ContainsAny` for each requires substance.
2. For rigid bodies (e.g. ice cubes, strawberries), we should check if the right proportion is in the blender when it's toggled on. E.g. if the recipe requires 1 strawberry + 2 ice cubes, then 2 strawberries + 4 ice cubes should also work.
3. If the proportion is right, we should generate the output substance that has the same rough volume of the input rigid bodies and substance. E.g. 2 strawberries + 4 ice cubes should produce more smoothie than 1 strawberry + 2 ice cubes. Right now, we are just setting Filled = True for the container, which can be a little unrealistic. 
4. If the proportion is wrong (including missing ingredient), a special substance ("garbage") should be outputted. 
Sorry this is a bit complicated, happy to discuss in person!

Reply:
these are good points. let's discuss in person once you get back

Reply:
a related implementation detail i think we may need to consider is how to pass relevant information from `condition` into `transition`, since `condition` may do some additional object filtering, e.g.: parsing which objects are actually inside the blender or not